community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/CreateNodePipe.scala

```diff
@@ -79,19 +75,52 @@ case class CreateNodePipe(src: Pipe, key: String, labels: Seq[LazyLabel], proper
     state.query.setLabelsOnNode(nodeId, labelIds.iterator)
   }
 
-  def planDescriptionWithoutCardinality = src.planDescription.andThen(this.id, "CreateNode", variables)
-
   def symbols = src.symbols.add(key, CTNode)
 
-  def withEstimatedCardinality(estimated: Double) = copy()(Some(estimated))
+  override def localEffects = if (labels.isEmpty)
+    Effects(CreatesAnyNode)
+  else
+    Effects(CreatesNodesWithLabels(labels.map(_.name).toSet))
+}
+
+case class CreateNodePipe(src: Pipe, key: String, labels: Seq[LazyLabel], properties: Option[Expression])(val estimatedCardinality: Option[Double] = None)
+                         (implicit pipeMonitor: PipeMonitor) extends BaseCreateNodePipe(src, key, labels, properties, pipeMonitor) {
+
+  override def withEstimatedCardinality(estimated: Double) = copy()(Some(estimated))
+
+  override def planDescriptionWithoutCardinality = src.planDescription.andThen(this.id, "CreateNode", variables)
 
   override def dup(sources: List[Pipe]): Pipe = {
     val (onlySource :: Nil) = sources
     CreateNodePipe(onlySource, key, labels, properties)(estimatedCardinality)
   }
 
-  override def localEffects = if (labels.isEmpty)
-    Effects(CreatesAnyNode)
-  else
-    Effects(CreatesNodesWithLabels(labels.map(_.name).toSet))
+  override def setProperty(nodeId: Long, key: String, value: Any, qtx: QueryContext) = {
+    //do not set properties for null values
+    if (value != null) {
+      val propertyKeyId = qtx.getOrCreatePropertyKeyId(key)
+      qtx.nodeOps.setProperty(nodeId, propertyKeyId, makeValueNeoSafe(value))
+    }
+  }
+}
+
+case class MergeCreateNodePipe(src: Pipe, key: String, labels: Seq[LazyLabel], properties: Option[Expression])(val estimatedCardinality: Option[Double] = None)
+                         (implicit pipeMonitor: PipeMonitor) extends BaseCreateNodePipe(src, key, labels, properties, pipeMonitor) {
+
+  override def withEstimatedCardinality(estimated: Double) = copy()(Some(estimated))
+
+  override def planDescriptionWithoutCardinality = src.planDescription.andThen(this.id, "MergeCreateNode", variables)
+
+  override def dup(sources: List[Pipe]): Pipe = {
+    val (onlySource :: Nil) = sources
+    MergeCreateNodePipe(onlySource, key, labels, properties)(estimatedCardinality)
+  }
+
+  override def setProperty(nodeId: Long, key: String, value: Any, qtx: QueryContext) = {
+    //merge cannot use null properties, since in that case the match part will not find the result of the create
```

Comment:
I would really like to see a document or note somewhere about why we needed to differentiate the null behaviour between create-node and merge-create-node.